io/js/src/main/scala/fs2/io/net/SocketPlatform.scala

```diff
@@ -53,7 +53,7 @@ private[net] trait SocketCompanionPlatform {
         }
       }
 
-  private[net] class AsyncSocket[F[_]](
+  private[net] case class AsyncSocket[F[_]](
       sock: facade.net.Socket,
       readStream: SuspendedStream[F, Byte]
```

Comment:
I've been thinking about this. Do we know what happens to this `SuspendedStream` after the `sock` is directly upgraded?

What I mean is that this `readStream` has already established listeners on the `sock`. So if events are firing on those listeners as `sock` is being used, but nobody is consuming from the `SuspendedStream`, then I am concerned that this is actually a memory leak.

Reply:
Good question.
As far as I can tell the listeners are all on control events, rather than data, i.e. they trigger changes of behaviour rather than push any data anywhere.

The read loop boils down to stream pull -> readable.read(), so if the stream isn't being consumed then `read() ` is never being called and there is no data to leak.

This also means that we're relying on the Javascript runtime to propagate any error or close events, which I think is a reasonable assumption, but we could add additional listeners for them if we feel that it's needed

Reply:
> As far as I can tell the listeners are all on control events, rather than data, i.e. they trigger changes of behaviour rather than push any data anywhere.

Right, sure. But those triggers still invoke callbacks, that typically put things in `Queue`s. If they are not consumed, they accumulate and leak.

Reply:
I don't think we're going to leak any data anywhere, but I guess we we'll end up with a channel full of `()`.

The only `Queue` that I can see in the process is `synchronous` so wont accumulate anything.

It seems like it ought to be possible for that channel not to be a channel since it's only acting as an indicator that something can be read or not (that is we need to communicate that something can be read, but duplicate events prior to a read are meaningless)

Reply:
> but I guess we will end up with a channel full of `()`.

Yes, sorry, this is what I meant by "memory leak". Not specifically a data leak, as in bytes of data.

> It seems like it ought to be possible for that channel not to be a channel since it's only acting as an indicator that something can be read or not (that is we need to communicate that something can be read, but duplicate events prior to a read are meaningless)

This is a great point. I wonder if we can replace it with a `Signal`.

Reply:
> The only `Queue` that I can see in the process is `synchronous` so wont accumulate anything.

Ah, this is a misconception. Unless whoever is pushing to the `synchronous` queue respect backpressure and stops trying to push, the attempts to push will queue up unboundedly.

Reply:
Yes sorry, I actually thought that we wouldn't leak anything (including `()`). Had to reread the Node docs to correct myself about how the readable event actually works (I thought it was only fired once, rather than repeatedly).

I'm happy to switch over to a `Signal` here but I wonder if it's worth it with #3348 now which should solve the issue as well. Don't know if you've got any preference for handling related PRs like these?


> Unless whoever is pushing to the synchronous queue respect backpressure

I admit I'd assumed that the internals of fs2 would get that right 😉

Reply:
> Had to reread the Node docs to correct myself about how the readable event actually works (I thought it was only fired once, rather than repeatedly).

It is _very_ confusing. While working on https://github.com/typelevel/fs2/pull/3348 I discovered that the `readable` event may be fired multiple times in a row (honestly it seemed a bit buggy) without `read()`s in between, which is why we needed a counter for readable events and not simply a boolean toggle ...

I think if you merge https://github.com/typelevel/fs2/pull/3348 into your PR, then my concerns about memory leaks would be addressed. But I need to think about this more. The "dangling" listeners are difficult to reason about.